content/05-mysql/10-reading-and-querying-data/05-optimizing-slow-queries.mdx

```diff
@@ -0,0 +1,446 @@
+
+The first considerations you'll need to make are what columns you'd like to index on.  If you regularly join on specific columns, it is important to make sure they are accounted for in your indexes.  Indexes are used primarily in row lookups, so join operations, where various rows in different tables are matched, are very expensive without index.
+
+You can create indexes with a single column if it fully encapsulates how you typically query the table.  You can also index on multiple columns to make the indexing more useful.  If indexing on multiple columns, the ordering is very important.  The first column should be the primary way to seek matches while any additional columns should help you refine those matches.
+
+For example, if you create an index for `first_name` and `last_name` for the table `person`, you'll almost certainly want to define the index with something like this:
+
+```sql
+CREATE INDEX last_and_first_names ON person (last_name, first_name);
+```
+
+Here, the `last_name` column is the first or left-most column in the index, meaning it will be fast looking up either `Alan Smith` or all `Smith` entries.  It won't be effective for looking up all instances of `Alan` as the `first_name` however, which is probably fine unless we anticipate querying only by first name often.
+
+While creating indexes can help you increase the read performance of your datasets, it is important to not over-index your tables.  Indexes come at a cost to write speed, as each time new values are recorded in the table, all of the indexes associated with it must also be updated.  Tables work best if their primary keys are well fitted with the querying patterns that'll be used against them as it allows you to benefit from strong index performance without needing a large number of supplemental indexes to accommodate different access patterns.
+
+### Limiting candidate rows
+
+Another way that you can speed up queries is by limiting the number of candidate and returned rows in your statements when possible.  This is a very context-specific recommendation, so it won't be appropriate in many cases, but if you can pare down the number of rows that MySQL has to evaluate, you can increase performance significantly.
+
+You can do this with the [`LIMIT` clause](https://dev.mysql.com/doc/refman/8.0/en/limit-optimization.html).  For example, to return only the five results, you can use a format like this:
+
+```sql
+SELECT * FROM <table> LIMIT 5
+```
+
+In many cases (depending on what other clauses are included in the query), limiting the number of rows returned can help MySQL short circuit the querying process, stopping sooner than it would normally need to.  This helps save time early on instead of executing over the entire dataset early on and then truncating the results later.
```

Comment:
```suggestion
In many cases (depending on what other clauses are included in the query), limiting the number of rows returned can help MySQL short circuit the querying process, stopping sooner than it would normally need to.  This helps save time early on instead of executing over the entire dataset and then truncating the results later.
```